interfaces/tests/file_int/file_int.c

```diff
@@ -0,0 +1,559 @@
+    char filename[] = "file_create_creates_new_file.txt";
+    ASSERT_IS_FALSE(check_file_exists(filename));
+
+    ///act
+    FILE_HANDLE file_handle = file_create(execution_engine, filename, NULL, NULL);
+
+    ///assert
+    ASSERT_IS_NOT_NULL(file_handle);
+    ASSERT_IS_TRUE(check_file_exists(filename));
+}
+
+/*Tests_SRS_FILE_43_003: [If a file with name full_file_name does not exist, file_create shall create a file with that name.]*/
+/*Tests_SRS_FILE_43_001: [file_create shall open the file named full_file_name for asynchronous operations and return its handle.]*/
+/*Tests_SRS_FILE_43_006: [file_destroy shall wait for all pending I / O operations to complete.]*/
+/*Tests_SRS_FILE_43_007: [file_destroy shall close the file handle handle.]*/
+/*Tests_SRS_FILE_43_014: [file_write_async shall enqueue a write request to write source's content to the position offset in the file. ]*/
+/*Tests_SRS_FILE_43_041: [If position + size is greater than the size of the file and the call to write is successfull, file_write_async shall grow the file to accomodate the write.]*/
+/*Tests_SRS_FILE_43_008: [file_write_async shall call user_call_back passing user_context and success depending on the success of the asynchronous write operation.]*/
+/*Tests_SRS_FILE_43_030: [file_write_async shall succeed and return FILE_WRITE_ASYNC_OK.]*/
+/*Tests_SRS_FILE_43_021: [file_read_async shall enqueue a read request to read handle's content at position offset and write it to destination. ]*/
+/*Tests_SRS_FILE_43_016: [file_read_async shall call user_callback passing user_context and success depending on the success of the asynchronous read operation.]*/
+/*Tests_SRS_FILE_43_031: [file_read_async shall succeed and return FILE_READ_ASYNC_OK.]*/
+TEST_FUNCTION(write_to_a_file_and_read_from_it)
+{
+    ///arrange
+    const int size = 5;
+
+    unsigned char source[] = "abcd";
+    WRITE_COMPLETE_CONTEXT write_context;
+    write_context.pre_callback_value = 41;
+    interlocked_exchange(&write_context.value, write_context.pre_callback_value);
+    write_context.post_callback_value = 42;
+
+    unsigned char destination[5];
+    READ_COMPLETE_CONTEXT read_context;
+    read_context.pre_callback_value = 43;
+    interlocked_exchange(&read_context.value, read_context.pre_callback_value);
+    read_context.post_callback_value = 44;
+
+    char filename[] = "write_to_a_file_and_read_from_it.txt";
+    ASSERT_IS_FALSE(check_file_exists(filename));
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    FILE_HANDLE file_handle = file_create(execution_engine, filename, NULL, NULL);
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT write_result = file_write_async(file_handle, source, size, 0, write_callback, &write_context);
+    wait_on_address_helper(&write_context.value, &write_context.pre_callback_value, UINT32_MAX);
+    
+    ///assert
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_OK, write_result);

```

Comment:
>ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_OK, write_result); [](start = 4, length = 77)

The synchronous check needs to happen before the wait, because the callback won't come if it failed. Check should be:

Check sync return
Wait for callback
Check callback status

Here and everywhere